WORKFLOW.md

```diff
@@ -0,0 +1,65 @@
+
+### Workflow
+
+- BucketRequest(namespace-scoped) can be created with reference to a BucketClass, so user will get a Bucket(cluster-scoped). The COSI driver will create a "backend bucket" in RGW for a Bucket Request. The owner is defined based on the BucketClass. So owner will be created for the first BucketRequest on the BucketClass.
+- BucketAccessRequest(namespace-scoped) can be created with a reference to a BucketAccessClass and BucketRequest, so user will get a BucketAccess(cluster-scoped).The COSI driver will create an S3 user with proper access.
+
+The BC and BAC are handled by the admin and user need to create the BR and BAR out of it. The app pod is given access to the BAR, both need to running on the same namespace. Even though BR is namespace-scoped, it can be accessed in different namespace known as `allowed namespaces`. So the BAR need to part of this namespace for accessing the BR.
+
+### COSI Driver
+The user credentials and endpoint need to passed as Kubernetes Secret to the driver. Using these details, the driver will perform bucket operations via S3 APIs and user operations via RADOS Gateway administrator operations (RGW admin ops) APIs. The following are the API's defined by the driver:
+- ProvisionerCreateBucket() - called upon create BucketRequest, the driver will create user with help of RGW admin ops API and then create a bucket using S3 API using that user.
+- ProvisionerDeleteBucket() - called upon delete BucketRequest, the driver will fetch BucketInfo using RGW admin ops API, then fetch the owner info, delete bucket via S3 API using the owner.
+- ProvisionerGrantBucketAccess() - called upon create BucketAccessRequest, create user with RGW admin ops API and attach the user to required bucket with help of bucket policy.
+- ProvisionerRevokeBucketAccess() - called upon delete BucketAccessRequest, revoke the permission on with user and then delete the user.
+
+```
+P.S Since there can be multiple users(BAR) attaching to the same bucket, it is not tested how much RGW scale with COSI.
+```
+
+### Consuming at application
+The app pod can refer BucketAccessRequest and Node adapter details as volume mount as follows:
+```yaml
+spec:
+  containers:
+      volumeMounts:
+        - name: cosi-secrets
+          mountPath: /data/cosi
+  volumes:
+  - name: cosi-secrets
+    csi:
+      driver: objectstorage.k8s.io
+      volumeAttributes:
+        bar-name: sample-bar
+        bar-namespace: default
+```
+The COSI adapter sidecar injects the Ceph driver's credentials into the application pod into the mount path given (/data/cosi in the above example).
+
+### Integrating with Rook
+The controller, node adapter and sidecar/driver need to be deployed. This design proposes that Rook deploy these components for a Rook-Ceph cluster. This design also proposes that Rook create a `CephObjectStoreUser` and refer to the credentials secret created by that controller.
```

Comment:
What are alternatives to Rook deploying these components? Why do we choose to have Rook deploy them instead of having a separate deployment method?

Reply:
Are there security risks with creating a `CephObjectStoreUser` for each request instead of creating the user direclty via API calls? Are there other reasons we choose to use the proposed method instead of creating things directly?

Reply:
I have mentioned deploying COSI Driver without using Rook in https://github.com/ceph/ceph-cosi/blob/master/README.md

`CephObjectStoreUser` is one-time creation for the cosi driver itself(like an admin user), not for the BARs handled by it